### PR TITLE
Add codebase.md

### PR DESCRIPTION
**Pull request by candour**
Describes the code base

- - -
This change is [Reviewable](https://reviewable.io/reviews/fluffle/sp0rkle/87)

### CODE REVIEW

File: codebase.md [hunk 1]
```diff
@@ -0,0 +1,110 @@
...
+    // inside the context, but usually, testing the underlying
+    // data-mangling functions is enough.
+}
+```
+
+See `drivers/factdriver/plugin_test.go` for real-world examples of exercising logic without descending into integration-test hell.
+
+## 5. Pro-Tips and Pitfalls
+
+- **The Global `bot`**: The `bot` package uses a global singleton. It's not "modern Go," but it works. Just be careful with state.
```
Comment (gundalow) [inline codebase.md:98]: Suggestions, define how to be careful. What false assumptions might humans & AI have if they are used to more modern Go

File: codebase.md [hunk 2]
```diff
@@ -0,0 +1,110 @@
...
+- **Data Loss (The Mongo Curse)**: We hate MongoDB. That's why we're migrating. If you find a data mismatch between Mongo and Bolt, the `db.Both` layer will log a warning. Listen to it.
+- **Regex is Your Friend (and Enemy)**: Much of sp0rkle's parsing relies on regex. Use `util/lexer.go` if you want to parse complex strings without descending into madness.
```
Comment (gundalow) [inline codebase.md:100]: Ensure unit tests are added for positive, negative and edge cases.

**Comment by gundalow** [comment]
Bots and humans won't read https://github.com/fluffle/sp0rkle/wiki so should that be copied into here (or maybe `docs/`directory? 

Should this he named AGENTS.md?

File: codebase.md [hunk 3]
```diff
@@ -0,0 +1,110 @@
...
+    // Register a raw handler for IRC events
+    bot.Handle(rawHandler, client.PRIVMSG)
+
+    // Register a rewriter (modifies outgoing text)
+    bot.Rewrite(myRewriter)
+}
+```
+
+### Step 3: Register in `main.go`
+Your driver won't do anything unless you import it in `main.go` and call its `Init()` function. Yes, this is manual. No, we don't have fancy auto-discovery.
```
Comment (fluffle) [inline codebase.md:41]: Lol'd.

<3 passive aggressive AI comments. Some sort of dependency injection framework would be nice to have I agree but nothing like that existed when I wrote sp0rkle and I sure as hell wasn't going to invent it :-D

File: codebase.md [hunk 4]
```diff
@@ -0,0 +1,110 @@
...
+```
+
+See `drivers/factdriver/plugin_test.go` for real-world examples of exercising logic without descending into integration-test hell.
+
+## 5. Pro-Tips and Pitfalls
+
+- **The Global `bot`**: The `bot` package uses a global singleton. It's not "modern Go," but it works. Just be careful with state.
+- **Data Loss (The Mongo Curse)**: We hate MongoDB. That's why we're migrating. If you find a data mismatch between Mongo and Bolt, the `db.Both` layer will log a warning. Listen to it.
+- **Regex is Your Friend (and Enemy)**: Much of sp0rkle's parsing relies on regex. Use `util/lexer.go` if you want to parse complex strings without descending into madness.
+- **Pollers**: If your feature needs to do something periodically (like checking an RSS feed), use `bot.Poll(myPoller)`.
```
Comment (fluffle) [inline codebase.md:101]: Pollers are shit, architecturally. They need to be completely decoupled from the connection lifecycle so they can be stopped and started arbitrarily. Also I think I got the API wrong. Needs some thought.

I wouldn't recommend using them in their current state even if they do just about solve the "recurring event" problem.

File: codebase.md [hunk 5]
```diff
@@ -0,0 +1,110 @@
...
+
+## 6. Development Environment
+
+1. Install Go and MongoDB (if you must).
+2. Run `go build` in the root.
+3. Run the bot: `./sp0rkle --servers irc.yournet.org --nick mybot --channels "#test"`.
+4. If you need a database backup, see `backup.sh`. If you lose your data because you didn't have a backup... well, that's the sp0rkle way.
```
Comment (fluffle) [inline codebase.md:108]: > well, that's the sp0rkle way.

sick burn

sp0rkle has not lost data in >20 years now :-)

**Review by fluffle — APPROVED** [review]
It's been too long for me to actually have proper context any more, but thanks for sending this :-)

File: codebase.md (hunk 3 above)
Comment (gundalow) [reply codebase.md:41]: The wording is based on your wording on the existing docs, I feel it's matched your style well.

**Review by gundalow — APPROVED** [review]
I think we merge this, then post https://github.com/fluffle/sp0rkle/pull/90 (which adds AGENTS.md) I'll move it to `docs/`.

Is codebase.md a new standard filename, I've not seen it before.

**Comment by fluffle** [comment]
Traditionally when reviews happen and comments are made the author _goes back to_ the PR and _addresses the comments_ before it's merged. Call me old-fashioned but I think that's still a valid thing to expect. I don't know how we make Jules do that -- perhaps I needed to "request changes" rather than just snarking -- but you had some valid requests and I think they were worth getting the AI to expend some more tokens on.

**Comment by fluffle** [comment]
Let's try:

@google-labs-jules please address @gundalow's requests to add additional suggestions on unit testing and "how to be careful".

**Comment by gundalow** [comment]
> Traditionally when reviews happen and comments are made the author _goes back to_ the PR a

Oh, I meant after this is merged. ie resolve all the above comments before merging.

>  I don't know how we make Jules do that
I think @candour needs to tell the bot to address all the review comments

**Comment by candour** [comment]
@google-labs-jules please address @gundalow's requests to add additional suggestions on unit testing and "how to be careful".

**Comment by candour** [comment]
@jules please respond to the PR comments
